mlir/lib/Dialect/MemRef/Transforms/ResolveShapedTypeResultDims.cpp

```diff
@@ -131,11 +132,25 @@ struct IterArgsToInitArgs : public OpRewritePattern<tensor::DimOp> {
     auto blockArg = dyn_cast<BlockArgument>(dimOp.getSource());
     if (!blockArg)
       return failure();
-    auto loopLikeOp =
-        dyn_cast<LoopLikeOpInterface>(blockArg.getParentBlock()->getParentOp());
-    if (!loopLikeOp)
+    // TODO: Enable this for loopLikeInterface. Restricting for scf.for
```

Comment:
fyi, for `scf.for` we already have this pattern here: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SCF/Transforms/LoopCanonicalization.cpp#L87

Reply:
I think it would make sense to move this pattern to the same file.

Reply:
I think that pattern is wrong for the reason why this is being restricted to forall only. The init arg shape can vary from iteration to iteration. There it is added as a canonicalization which is even worse.

Reply:
This pattern has an additional `isShapePreserving` check that makes it safe. Btw, that check is quite conservative, it could be improved by checking for `DestinationStyleOpInterface` instead of hard-coding a few ops in the analysis.

Reply:
@MaheshRavishankar @matthias-springer I'll be moving the pattern from LoopCanonicalization to resolveShapedTypeResultDims and updating that to process scf.forall. That sounds reasonable to me.

Reply:
I'll be merging this and refactoring in upcoming PRs.